### PR TITLE
[inductor] convert layout of conv weight ahead of time for inference

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103642

This PR handles inference. Will do similar thing for training later.

Some manual testing results shows this can improve inference perf by 2-3% (absolute improvement not relative one).
- convmixer: 4.285x -> 4.309x
- resnet50: 2.170x -> 2.203x

The PR is built upon freezing. Since without freezing, the weight input for a conv node may not be a parameter directly but be the output of precision converting ops. It's so much easier to implement this PR after freezing.

Commands
```
TORCHINDUCTOR_FREEZING=1 python benchmarks/dynamo/timm_models.py --backend inductor --amp --performance --only convmixer_768_32 --inference  
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

### CODE REVIEW

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -679,6 +679,8 @@ def fw_compiler_freezing(
     # partition_fn won't be called
     joint_graph_passes(aot_autograd_model)
 
+    convert_conv_weights_to_channels_last(aot_autograd_model, aot_example_inputs)
+
```
Comment (XiaobingSuper) [inline torch/_inductor/compile_fx.py:683]: I have a question about it, for convolution, if input or weight has one channels_last, the output format will be channels_last, if we do a format change there, do we need to update the  ```node.meta``` info(there have such infos: {"val": FakeTensor, "tensor_meta": TensorMetadata})?

Comment (shunting314) [reply torch/_inductor/compile_fx.py:683]: yea, we need change both layout for compile time input (in `Node.meta['val']`) and layout for runtime input to make them consistent.

BTW, this PR is WIP. I have some related comments in another file but have not implement that yet. I'll remove 'WIP' tag in the title when the PR is ready.

File: torch/_inductor/freezing.py [hunk 2]
```diff
@@ -224,3 +230,36 @@ def discard_traced_gm_params(mod):
...
+            with no_dispatch():
+                cl_param_tensor = param_tensor.to(memory_format=torch.channels_last)
+                if isinstance(param_tensor, nn.Parameter):
+                    cl_param_tensor = nn.Parameter(cl_param_tensor)
+            if cl_param_tensor is not param_tensor:
+                setattr(gm, weight_node.target, cl_param_tensor)
+
+                # Even though inductor does not use meta['val'] or meta['tensor_meta']
+                # for get_attr node, we still update them to be consistent.
+                weight_node.meta["val"] = weight_node.meta["val"].to(
```
Comment (XiaobingSuper) [inline torch/_inductor/freezing.py:260]: We may not use ```meta['val']``` or ```meta['tensor_meta']``` for the get_attr node, we change the weight format here, the convolution node infos may also need to be changed or we don't need to care about it?

Comment (shunting314) [reply torch/_inductor/freezing.py:260]: I think if we want the meta data in every node of the graph to be consistent with the layout conversion, we need do a full pass thru the graph and update potentially every reached node's layout. But inductor may make different decision to the saved layout (in meta['val']) anyways, so making meta['val'] for intermediate nodes up to date sounds like an over-kill.  Leave them reflecting the original eager layout pretending no layout conversion looks not too bad.

File: torch/_inductor/freezing.py [hunk 3]
```diff
@@ -224,3 +230,36 @@ def discard_traced_gm_params(mod):
...
+    for conv in convs:
+        weight_node = conv.args[1]
+        # is a constant tensor
+        if weight_node.op == "get_attr":
+            param_tensor = getattr(gm, weight_node.target)
+            if len(param_tensor.shape) != 4:
+                # not a 4d tensor, skip
+                continue
+            with no_dispatch():
+                cl_param_tensor = param_tensor.to(memory_format=torch.channels_last)
```
Comment (eellison) [inline torch/_inductor/freezing.py:252]: Could we convert the tensors to channels last by transformation in the graph ? This is going to compose better with the rest of the freezing stack... In particular, to do real-time weight updates, we are going to need to record all of the transformations we do on weights, and then reapply them. If the channels-last conversion occurs in the graph, it will occur naturally with the other transforms, like mm weight padding

Sorry this is kind of a new wrinkle (real-time updates) I just heard about.

File: torch/_inductor/freezing.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/freezing.py:260]: It might be easier to just inset the weight conversions, and then retrace to get the updated .meta['val']. 

Changing weights to channels last could affect more than just the weight_node meta, other nodes as well.

**Review by eellison — COMMENTED** [review]
Cool ! A few comments. let me know what you think

File: torch/_inductor/freezing.py (hunk 2 above)
Comment (shunting314) [reply torch/_inductor/freezing.py:260]: Hmm, I still think we should not do it ..

Layout optimization is something not visible to user hopefully. Most of the time .meta['val'] represents the eager stride *without* layout optimization. If we retrace with CL inputs,  then what we end up have in .meta['val'] are those strides *with* layout optimizations.  For intermediate tensors, this is fine since they are not user visible, but it's bad for output tensors. We have been take care to maintain the same stride as eager for output tensors, we probably should not break it.

File: torch/_inductor/freezing.py (hunk 3 above)
Comment (shunting314) [reply torch/_inductor/freezing.py:252]: No worries. Just to clarify, so you suggest to use a fx.Transformer to do what I'm doing here?

Comment (eellison) [reply torch/_inductor/freezing.py:252]: We would be inserting `torch.ops.aten.clone.default(weight, memory_format = torch.channels_last)` nodes in the graph on the weights. You could use a fx.Transformer here but i think just iterating over the graph is fine.

Comment (shunting314) [reply torch/_inductor/freezing.py:252]: I see. Do also need to reorder the freezing pass and convert_conv_weights_to_channels_last pass?

Currently convert_conv_weights_to_channels_last happens right after freezing. But if we are add the layout conversion nodes (`aten.clone`) to the graph, we should do convert_conv_weights_to_channels_last before the freezing pass? Otherwise those aten.clone node gonna appear at runtime.

Comment (eellison) [reply torch/_inductor/freezing.py:252]: Yea, that's correct. We would need to insert it before the freeze call here https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L684, on the aot_autograd graph

**Comment by jansel** [comment]
At a glance this looks good to me.  I'll let @eellison do the final review.

**Comment by shunting314** [comment]
@eellison @jansel  this is more complex then I initially thought. I just found one issue. We convert convolution weights to channels last (either change GraphModule attribute directly or add aten.clone nodes to the graph to do the layout convolution). Later on in `compile_fx_inner` when we do FakeTensorProp, the changed strides of the conv weight will be propagated thru the graph. After the propagation, a previously contiguous *output* tensor in eager may become a channels last tensor. 

I think to avoid that and make sure we don't change output tensor's layout, we need
1. collect all output tensors layout before the pass added by this PR. Those represent the eager layout we need for output tensors
2. perform the pass added by this PR
3. explicitly add aten.clone node in the end of the graph to convert each of the output tensor's stride to those collected in step 1.

Another variant of step3 is, we collect all the output tensor's stride after performing step2 (Need do an extra FakeTensorProp) and apply step3 only to those output tensors whose stride get changed. But this may not be necessary since I assume adding redundant no-op aten.clone node is fine. We can remove those no-op aten.clone nodes from the graph in a post_grad pass. (maybe we already did).

Wdyt?

File: torch/_inductor/freezing.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/freezing.py:252]: Wrote up the issue here: https://github.com/pytorch/pytorch/issues/103990

**Comment by eellison** [comment]
@shunting314 for output nodes which are contiguous, or channels-last, I think it makes sense to insert conversion nodes on the outputs that would maintain the original striding. For non-dense outputs, we don't preserve their stride now, [anyway](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/graph.py#L668C1-L669). 

 I think the easiest way to do that would be to insert `.to(layout=...)` nodes at the on the definition of these outputs, and then re-trace through the graph. If you trace again, no-op `.to` nodes will not remain in the graph. 
 
 The other alternative is to as you said to call FakeTensorProp, and then manually convert the `.to` nodes to either be clones or removed. 

What do you think ?

**Comment by shunting314** [comment]
> I think the easiest way to do that would be to insert `.to(layout=...)` nodes at the on the definition of these outputs, and then re-trace through the graph. If you trace again, no-op `.to` nodes will not remain in the graph.
> 
> The other alternative is to as you said to call FakeTensorProp, and then manually convert the `.to` nodes to either be clones or removed.
> 
> What do you think ?

Will re-tracing increase compilation time since the whole graph needs to be re-created?

**Comment by eellison** [comment]
In this case, we'd only be re-tracing for inference. If you look at our compilation time there it's much lower. I don't know exactly how much overhead one fx trace is. We could benchmark and decide from there.

File: test/inductor/test_inductor_freezing.py [hunk 4]
```diff
@@ -229,6 +232,140 @@ def foo(mod, inp):
...
+            nonlocal num_same_stride, num_diff_stride
+            input_tensor.realize()
+            if tuple(input_tensor.get_stride()) == tuple(stride):
+                num_same_stride += 1
+            else:
+                num_diff_stride += 1
+            return orig_inductor_force_stride(input_tensor, stride)
+
+        @contextlib.contextmanager
+        def mock_inductor_force_stride():
```
Comment (eellison) [inline test/inductor/test_inductor_freezing.py:325]: The tests here are great, but seem like a lot of work. I wonder how much of this could have been done with `run_and_get_code`, I think a lot. Or similarly, if we should add other lightweight mechanisms to access various parts of the compilation stack via tests. Like, `run_and_get_inductor_graph`, which would dump out the inductor graph.

**Review by eellison — COMMENTED** [review]
Very cool ! A number of smaller questions and just one question on whether we want to reuse [find_nodes_prefer_channels_last](https://github.com/pytorch/pytorch/blob/3e674b75b1d307d363ef5e626e098c65b5016d6f/torch/_inductor/graph.py#L314). 

Addressing that doesn't have to be land-blocking, up to you.

File: torch/_inductor/freezing.py [hunk 5]
```diff
@@ -241,3 +245,60 @@ def discard_traced_gm_params(mod):
...
+            # n.replace_all_uses_with(new_node)
+            # since it will replace the usage of n in new_node itself.
+            output_node.replace_input_with(n, new_node)
+
+    gm.graph.lint()
+    gm.recompile()
+
+
+@dynamo_timed
+def convert_conv_weights_to_channels_last(gm):
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/freezing.py:282]: may be not. `find_nodes_prefer_channels_last` is used to make sure we keep channels last format for saved 4d activations. It's not supposed to affect the handling of weights on purpose.

Regarding handling of weights, right now we only convert convolution weights to channels last ahead of time. I think that's probably is good enough. Weights for other ops like batch_norm / layer_norm are not 4d tensors. So there is no channels last format for them.  But maybe there is better layout than contiguous for those weights. I'm not sure.

File: torch/_inductor/freezing.py [hunk 6]
```diff
@@ -241,3 +245,60 @@ def discard_traced_gm_params(mod):
...
+
+def enforce_output_layout(gm):
+    """
+    Make sure the output node's layout does not change due to compiler optimizations
+    by adding aten.as_strided nodes with the expected strides.
+
+    Only used for inference so we can assume all graph outputs are model outputs.
+    """
+    *_, output_node = gm.graph.nodes
+    out_list, spec = pytree.tree_flatten(output_node.args)
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/freezing.py:258]: Is this guaranteed by dynamo or AOTAutograd? I remember we have some unit test to call inductor with graphs not captured by dynamo.

File: torch/_inductor/freezing.py [hunk 7]
```diff
@@ -241,3 +245,60 @@ def discard_traced_gm_params(mod):
...
+    with gm.graph.inserting_before(output_node):
+        for n in out_list:
+            if not isinstance(
+                n.meta["val"], torch.Tensor
+            ) or not torch._prims_common.is_non_overlapping_and_dense(n.meta["val"]):
+                continue
+
+            # add a node to enforce eager layout
+            ft = n.meta["val"]
+            new_node = gm.graph.call_function(
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/freezing.py:268]: Hm, is this very rare that we have a single tensor appear multiple times in the output?

I can add the cache if that happens common enough. Otherwise the next day when I or someone else reads the code, people may wondering why such a cache is necessary.

File: torch/_inductor/freezing.py [hunk 8]
```diff
@@ -241,3 +245,60 @@ def discard_traced_gm_params(mod):
...
+                prims.inductor_force_stride.default, (n, ft.stride())
+            )
+
+            # can not call
+            # n.replace_all_uses_with(new_node)
+            # since it will replace the usage of n in new_node itself.
+            output_node.replace_input_with(n, new_node)
+
+    gm.graph.lint()
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/freezing.py:277]: I agree this is only helpfully for debugging. But if we don't call it here (or control via a config and disable by default), then when people print the code of the graph module later, they may see stale code and get surprised. They can see up-to-date code by enabling the config, but people may not know there is such a config in the first place. That seems to be not good ..  Let me know if this is not a valid concern

File: test/inductor/test_inductor_freezing.py (hunk 4 above)
Comment (shunting314) [reply test/inductor/test_inductor_freezing.py:325]: `run_and_get_code` is great but may not help much here because it's not easy to decide if we have redundant clone by looking at the generated triton code.

`run_and_get_inductor_graph` can be a good alternative to do some assertion in fx Graph level. But if there is no objection, I still want to do the assertion in lowering (which happens laster compared to `run_and_get_inductor_graph`). 

I can make it a common API to override a lowering function. This way, code used in this test can be reused in other places .

File: torch/_inductor/freezing.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/freezing.py]: Seems it's guaranteed by `fx.Graph.output`

File: torch/_inductor/compile_fx.py [hunk 9]
```diff
@@ -729,11 +755,17 @@ def fw_compiler_freezing(
...
+    from torch._inductor.freezing import convert_conv_weights_to_channels_last, freeze
 
     # partition_fn won't be called
     joint_graph_passes(aot_autograd_model)
 
+    layout_opt = GraphLowering.decide_layout_opt(aot_autograd_model)
+    if layout_opt:
+        # make sure meta['val'] is properly setup
+        fake_tensor_prop(aot_autograd_model, aot_example_inputs, True)
+        convert_conv_weights_to_channels_last(aot_autograd_model)
```
Comment (eellison) [inline torch/_inductor/compile_fx.py:767]: I would have thought we needed the fake_tensor_prop after, but not before. I do think we need it after or the meta's will be incorrect

File: torch/_inductor/freezing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/freezing.py:277]: We can address this in a follow up, and more selectively recompile throughout inductor. Maybe when logging is set, debug is set, or pdb is set.

Or add a flag to GraphModule that will recompile upon printing.

File: torch/_inductor/lowering.py [hunk 10]
```diff
@@ -1470,6 +1470,12 @@ def randn(*args, **kwargs):
     raise AssertionError("should have been handled in replace_random.py")
 
 
+@register_lowering(inductor_prims.force_stride, type_promotion_kind=None)
```
Comment (eellison) [inline torch/_inductor/lowering.py:1473]: nit: maybe would be more faithful to this operator to make this something like `inductor_force_stride_order`, or, `inductor_force_stride_permutation` since we arent supporting generic strides here. 

not really important....

**Review by eellison — APPROVED** [review]
sweet !!

🚢 🚢 🚢 🚢 🚢 🚢 🚢
 

You might need to rebase past https://github.com/pytorch/pytorch/pull/100653 - I thinkkkk we should be good but might require small change

File: torch/_inductor/compile_fx.py (hunk 9 above)
Comment (shunting314) [reply torch/_inductor/compile_fx.py:767]: Run a FakeTensorProp before the conversion to make sure meta['val'] is properly setup. I'm worried that they may not already be setup.

But if we are sure FakeTensorProp has already been called before, we can save this call.

Comment (eellison) [reply torch/_inductor/compile_fx.py:784]: I think we should add it again, since now the existing meta's will be incorrect, since we've changed striding
